.github/workflows/branch.yml

```diff
@@ -52,28 +57,38 @@ jobs:
           python -m pip install --upgrade pip
           pip install -r requirements.txt
       - name: Test examples
-        run: ipython ${NOTEBOOK_TO_TEST}
+        run: |
+          echo "Running ${{ matrix.notebooks }}"
+          ipython ${{ matrix.notebooks }}
+  get-changed-scripts:
+    runs-on: ubuntu-latest
+    outputs:
+      changed_files: ${{ steps.output.outputs.changed_files }}
+    steps:
+    - uses: actions/checkout@v2
+    - name: Get changed files
+      id: changed-files
+      uses: tj-actions/changed-files@v23.1
+      with:
+        separator: "', '"
+        files: |
+          **/*.py
+    - name: Set output
+      id: output
+      if: steps.changed-files.outputs.any_changed == 'true'
+      run: |
+        echo "${{ steps.changed-files.outputs.all_changed_files }}"
+        echo "::set-output name=changed_files::[ '${{ steps.changed-files.outputs.all_changed_files }}' ]"
   test-scripts:
+    needs: get-changed-scripts
+    if: ${{ needs.get-changed-scripts.outputs.changed_files }}
     runs-on: ${{ matrix.os }}
     strategy:
       fail-fast: false
       matrix:
-        os: [ubuntu-latest]
-        python-version: [3.7]
-        scripts:
-          - how-to-guides/hello-world/scripts
-          - how-to-guides/hello-world/scripts
-          - how-to-guides/how-it-works/scripts
-          - how-to-guides/monitor-ml-runs/scripts
-          - how-to-guides/organize-ml-experimentation/scripts
-          - integrations-and-supported-tools/matplotlib/scripts
-          - integrations-and-supported-tools/sklearn/scripts
-          - integrations-and-supported-tools/tensorflow-keras/scripts
-          - integrations-and-supported-tools/catalyst/scripts
-          - integrations-and-supported-tools/lightgbm/scripts
-          - integrations-and-supported-tools/xgboost/scripts
-          - integrations-and-supported-tools/pytorch-lightning/scripts
-          - integrations-and-supported-tools/fastai/scripts
+        os: [ ubuntu-latest, windows-latest, macos-latest ]
+        python-version: [ '3.7', '3.8', '3.9', '3.10' ]
```

Comment:
for now let's do only for ubuntu and py 3.7, try how it works then expand

Reply:
I've actually tested on all combinations for both scripts and notebooks (hence the 50+ commits).
And since these will run only on the changed files, I think we can afford to include everything.
Worst case, we can have only the edge-cases (run only on python 3.7 and 3.10). This will half the number of tests.

What do you think?

Reply:
@SiddhantSadangi ack